okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,238 @@
+                                            .build();
+                                    client.newCallWithMutableState(
+                                            redirectedRequest, backoffStrategy, maxNumRelocations - 1)
+                                            .enqueue(callback);
+                                }
+                            }
+                            return null;
+                        }
+
+                        @Override
+                        public Void visit(QosException.Unavailable exception) {
+                            Optional<Duration> backoff = backoffStrategy.nextBackoff();
+                            if (!backoff.isPresent()) {
+                                log.debug("Max number of retries exceeded, failing call");
+                                callback.onFailure(call, new IOException("Failed to complete the request due to a "
+                                        + "server-side QoS condition: 503", exception));
+                            } else {
+                                log.debug("Rescheduling call after backoff",
+                                        SafeArg.of("backoffMillis", backoff.get().toMillis()));
+                                scheduleExecution(callback, backoff.get());
+                            }
+                            return null;
+                        }
+                    });
+                    return;
+                }
+
+                // Handle responses that correspond to RemoteExceptions / SerializableErrors
+                Optional<RemoteException> httpError = remoteExceptionHandler.handle(response);
+                if (httpError.isPresent()) {
+                    callback.onFailure(call, new IoRemoteException(httpError.get()));
+                    return;
+                }
+
+                // Catch-all: handle all other responses
+                Optional<IOException> ioException = catchAllExceptionHandler.handle(response);
+                if (ioException.isPresent()) {
+                    callback.onFailure(call, ioException.get());
+                    return;
+                }
+
+                callback.onFailure(call, new IOException("Failed to handle request, this is an http-remoting bug."));
+            }
+        });
+    }
+
+    @Override
+    public Response execute() throws IOException {
+        SettableFuture<Response> future = SettableFuture.create();
+        enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException exception) {
+                future.setException(exception);
+            }
+
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                future.set(response);
+            }
+        });
+
+        try {
+            return future.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.interrupted();
```

Comment:
`Thread.interrupted();` just returns a boolean - not sure this is doing what you want:

https://github.com/palantir/http-remoting/pull/587/files#diff-07eb3e409f7215a8a8b71594d6a0b47bR52

Reply:
ouch. thanks, fixed.